### PR TITLE
Support non-public cloud environments in the Azure Storage Queue and Azure Storage Blob scalers

### PR DESCRIPTION
**Pull request by amirschw**
Introduce an optional `cloud` parameter to the Azure Storage Queue and Azure Storage Blob scalers to support non-public Azure clouds. `cloud` can be set to one of the known Azure clouds, or to `Private`. When set to `Private`, the `endpointSuffix` parameter must also be provided. Otherwise, the endpoint suffix is inferred from the cloud environment.

PR for updating the documentation: https://github.com/kedacore/keda-docs/pull/463

Fixes (partially) #1285.

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [ ] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Changelog has been updated

Signed-off-by: amirschw <24677563+amirschw@users.noreply.github.com>

### CODE REVIEW

**Comment by tomkerkhove** [comment]
.NET has an enum that allows you to define the type and get the suffix automatically, isn't there a similar feature in Go?

**Comment by amirschw** [comment]
Thanks for the comment @tomkerkhove, I couldn't find something quite like that in the Azure SDK for Go, but did implement a similar method for the `StorageEndpointType` type. Please let me know what you think.

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,6 +19,8 @@
 
 ## Unreleased
 
+- Support non-public cloud environments in the Azure Storage Queue and Azure Storage Blob scalers ([#1863](https://github.com/kedacore/keda/pull/1863))
+
 ### New
 
 - TODO ([#XXX](https://github.com/kedacore/keda/pull/XXX))
```
Comment (tomkerkhove) [inline CHANGELOG.md:26]:
```suggestion

### New

- Support non-public cloud environments in the Azure Storage Queue and Azure Storage Blob scalers ([#1863](https://github.com/kedacore/keda/pull/1863))
```

**Comment by tomkerkhove** [comment]
Fine by me, I've opened https://github.com/Azure/azure-storage-queue-go/issues/25 to see if this can be simplified.

**Comment by amirschw** [comment]
Oh, now I get that you meant the type of cloud. This could have helped to simplify things if it wasn't for some air-gapped cloud environments, where the endpoint suffix cannot be made public.  For those environments, I believe that users will have to provide the endpoint suffix themselves.

**Comment by tomkerkhove** [comment]
I think the sweet spot is both where `cloud` defaults to the public cloud and they can configure the others and we determine the endpoint suffix. 

If it's stack or airgapped, then they use `Private` and have to configure `endpointSuffix`.

What do you think?

**Comment by amirschw** [comment]
I like that, thanks! I'll work on the necessary changes: adding an optional `cloud` config, and make `endpointSuffix` required only when `cloud` is set to `Private` (or `Other`?).

**Comment by tomkerkhove** [comment]
Sounds good to me! I would personally go with Private but I'm open to anything

**Comment by tomkerkhove** [comment]
Is there a reason why we are changing the Azure Event Hubs scaler? Because not all parts seem to be aligned?

**Comment by amirschw** [comment]
The Event Hub scaler uses `ParseAzureStorageBlobConnection` from azure_storage.go for getting checkpoints from blob storage. It does so using a connection string, but since I added the `endpointSuffix` argument to the function, I just had to pass an empty string there. Hope that answers your question.

**Comment by amirschw** [comment]
While this PR partially fixes #1285, There is still work to be done to support Government and other non-public clouds in the rest of the Azure scalers (Azure Event Bub, Azure Log Analytics, Azure Monitor, Azure Service Bus). I'm not sure I'll have the capacity to add this to all of them, but will try to find time to set up another PR (at least for the Service Bus scaler which is used by our team :)).
Maybe it's a good idea to open separate issues for each of the scalers to track this. Otherwise the original issue can probably be used for tracking all of them.

**Comment by tomkerkhove** [comment]
Would you mind resolving the merge conflict please? Then @zroubalik can do a review.

**Comment by amirschw** [comment]
Resolved the merge conflicts and ready for review 😄

**Review by zroubalik — APPROVED** [review]
LGTM, thanks @amirschw!

**Pull request merged**